packages/components/src/size-control/index.tsx

```diff
@@ -0,0 +1,127 @@
+/**
+ * WordPress dependencies
+ */
+import { __ } from '@wordpress/i18n';
+
+/**
+ * Internal dependencies
+ */
+import { Button } from '../button';
+import RangeControl from '../range-control';
+import { Flex, FlexItem } from '../flex';
+import type { SizeControlProps } from './types';
+import {
+	default as UnitControl,
+	parseQuantityAndUnitFromRawValue,
+	useCustomUnits,
+} from '../unit-control';
+
+import { Spacer } from '../spacer';
+
+export const DEFAULT_UNITS = [ 'px', 'em', 'rem', 'vw', 'vh' ];
+
+function SizeControl( props: SizeControlProps ) {
```

Comment:
@WordPress/gutenberg-components how do we feel about the new `SizeControl` component being introduced as a separate component (a private API to start with)?

Reply:
As @jasmussen [said](https://github.com/WordPress/gutenberg/pull/62328#issuecomment-2178531035), a big change in core components like this should be discussed in a separate issue to figure out the best approach. And the "why" of the change is not immediately clear to me from this PR — we may be able to find other approaches once the "why" is more clear.

Reply:
IMO, `SizeControl` shouldn't live in the `@wordpress/components` package, since it feels too tailored to a specific need of the editor.

We could instead discuss the need for a `UnitControl` + `RangeControl` component, which was already initially discussed in https://github.com/WordPress/gutenberg/issues/40507 and then abandoned for lack of priority.

Reply:
Thanks everyone for your input.

> And the "why" of the change is not immediately clear to me from this PR — we may be able to find other approaches once the "why" is more clear.

@mirka I submitted the component in a new PR: https://github.com/WordPress/gutenberg/pull/62866

> IMO, SizeControl shouldn't live in the @wordpress/components package, since it feels too tailored to a specific need of the editor.

@ciampo What does it mean 'tailored for a specific need of the editor' in this context?
I don't see how SizeControl is more tailored 'for a specific need of the editor' than the FontSizePicker. Could you explain, please?

Reply:
> I don't see how SizeControl is more tailored 'for a specific need of the editor' than the FontSizePicker. Could you explain, please?

That is true, and in fact I also think that `FontSizePicker` shouldn't be in the `@wordpress/components` package either for the same reason.

Reply:
Looking at the code, it seems to me that the reason the SizeControl component was added is to make several sets of UI common.

If that is the only purpose, couldn't the SizeControl component be placed as an internal component in the global styles directory rather than in the component package? In other words, `packages/edit-site/src/components/global-styles/font-sizes/size-control.js`.

In the future, if something like the `SliderControl` described in #40507 is added to the components package, we can easily replace the internal component with the official one.

Reply:
@t-hamano that would be my suggestion too — @mirka , what you think?

Reply:
> the reason the SizeControl component was added is to make several sets of UI common. If that is the only purpose, couldn't the SizeControl component be placed as an internal component in the global styles directory rather than in the component package?

Isn't that the reason behind all the components? I don't get why certain things should be components and certain not. The criteria doesn't seem to be clear at least for me.

Reply:
I think the criteria for adding new components to the components package are defined [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md).

The components package is used by many developers. The reason I hesitate to add new components to a component package is that once a new component is exposed, it cannot be easily removed or modified, and requires maintenance work.

I feel that the versatility of the `SizeControl` component has not been fully evaluated so far. I think we can implement it as an internal component and move forward with this PR until the discussion progresses.

Reply:
I'm not enthusiastic about duplicating all the SizeControl logic/styles and making them live in wp-edit-site though. I wouldn't say it's a trivial component.

The possibilities I see are:

- Rework the `disableCustomFontSizes` prop in FontSizePicker so that it allows for custom font sizes only (which I believe would be the equivalent of the proposed SizeControl). For example we could deprecate `disableCustomFontSizes` and make a new enum prop with three options (e.g. `"custom" | "predefined" | "both"`).
- Extract to a component as proposed in #62866. For context, we do have some [precedent](https://github.com/WordPress/gutenberg/issues/52734#issuecomment-2038252660) in DateTimePicker, and I have seen a somewhat similar proposal for SizeControl already (#61180). We're definitely seeing devs needing more modularity like this. In this case, I think I would prefer exporting it as a subcomponent of FontSizePicker (e.g. `FontSizePicker.CustomInput`) rather than a completely separate component.

Reply:
> I think the criteria for adding new components to the components package are defined [here](https://github.com/WordPress/gutenberg/blob/trunk/packages/components/CONTRIBUTING.md?rgh-link-date=2024-06-26T14%3A38%3A55Z).

Yes, thanks, but I had already read that before I commented. Still, I don't see how SizeControl differs from other components available in the package and why it should not exist in the components package.

----

Anyway, I am not available to continue pursuing the component path anymore. It isn't the main objective of this PR to add a new component but to provide a solution for users wanting to manage [font size](https://github.com/WordPress/gutenberg/issues/61987) and [spacing](https://github.com/WordPress/gutenberg/issues/61986) presets. Creating the component was a side task motivated by the desire to avoid repeating code from the font size picker, but that's taking much more time than I anticipated and blocking the original objective. I'll add the UI for this inside `packages/edit-site/global-styles/size-control` and maintain the code as short as possible to make the preset management work.

If there's a need for a new component can be added later as a follow-up.

Reply:
I think that this suggestion from @mirka hits the sweet spot between reducing the amount of new code introduced, future maintenance need and speed of iteration:

> Rework the `disableCustomFontSizes` prop in `FontSizePicker` so that it allows for custom font sizes only (which I believe would be the equivalent of the proposed SizeControl). For example we could deprecate disableCustomFontSizes and make a new enum prop with three options (e.g. "custom" | "predefined" | "both").

I'll see if I can spin up a PR with the necessary changes, so that this PR can use directly `FontSizePicker` — would that be ok @matiasbenedetto @mirka @t-hamano ?

Reply:
That should be a follow-up. The component has already consumed too much time for this implementation.

Reply:
@matiasbenedetto What are the requirements for #61986 that cannot be addressed by modifying FontSizePicker? Is there some logic in FontSizePicker that makes it unsuitable as a more generic sizing control, or is it a terminology issue?

Reply:
At a first glance I see 3 issues, that make me think that adding a 'mode' prop to determine the component to determine what it renders:
- FontSizePicker control only handle a small subset of css units.
- FontSizePicker naming.
- Adding this complexity doesn't seem to add much value to the FontSizePicker component and degrades the developer experience and it's more prone to errors.

Still, I haven't started working on the implementation of https://github.com/WordPress/gutenberg/issues/61986 yet. So I may find other reasons on the way.